File: Ivan-fullstack-dev/assets/frontend-template/src/api/menus.ts
import request from './request'

export interface Menu {
  id: number
  name: string
  /** 节点类型：1=目录 2=菜单(页面) 3=按钮（统一权限树） */
  menuType: number
  path?: string
  icon?: string
  parentId?: number | null
  fullPath?: string
  level: number
  sort: number
  isVisible?: boolean
  isEnabled?: boolean
  /** 权限编码（按钮节点，两段式如 users:edit） */
  permissionCode?: string | null
  /** 绑定的后端接口（按钮节点，如 UserController.Update） */
  controllerAction?: string | null
  remark?: string | null
  createTime?: string
}

export interface MenuTreeNode {
  id: number
  name: string
  menuType: number
  path?: string
  icon?: string
  parentId?: number | null
  fullPath?: string
  level: number
  sort: number
  isVisible: boolean
  isEnabled: boolean
  permissionCode?: string | null
  controllerAction?: string | null
  remark?: string | null
  hasChildren: boolean
  children: MenuTreeNode[]
}

export interface MenuCreateRequest {
  name: string
  /** 节点类型：1=目录 2=菜单 3=按钮（默认 2） */
  menuType: number
  path?: string
  icon?: string
  parentId?: number | null
  sort?: number
  isVisible?: boolean
  isEnabled?: boolean
  permissionCode?: string
  controllerAction?: string
  remark?: string
}

export interface MenuUpdateRequest {
  name: string
  menuType: number
  path?: string
  icon?: string
  parentId?: number | null
  sort: number
  isVisible?: boolean
  isEnabled?: boolean
  permissionCode?: string
  controllerAction?: string
  remark?: string
}

export interface MenuSortRequest {
  id: number
  targetParentId?: number | null
  beforeId?: number | null
  afterId?: number | null
}

/** 获取全部菜单（平铺列表） */
export function getMenusAll() {
  return request.get<unknown, Menu[]>('/menus')
}

/** 获取菜单树（全量，含 children，统一树：目录/菜单/按钮） */
export function getMenuTree() {
  return request.get<unknown, MenuTreeNode[]>('/menus/tree')
}

/** 获取前端侧边栏菜单树（仅启用+可见节点） */
export function getFrontendTree() {
  return request.get<unknown, MenuTreeNode[]>('/menus/frontend-tree')
}

/** 懒加载：获取指定节点的子节点 */
export function getMenuChildren(parentId: number) {
  return request.get<unknown, MenuTreeNode[]>(`/menus/${parentId}/children`)
}

/** 获取单个菜单 */
export function getMenuById(id: number) {
  return request.get<unknown, Menu>(`/menus/${id}`)
}

/** 按路径查询 */
export function getMenuByPath(path: string) {
  return request.get<unknown, Menu>('/menus/by-path', { params: { path } })
}

/** 获取指定节点的所有子孙节点 */
export function getMenuDescendants(id: number) {
  return request.get<unknown, Menu[]>(`/menus/${id}/descendants`)
}

/** 新增菜单 */
export function createMenu(data: MenuCreateRequest) {
  return request.post<unknown, number>('/menus', data)
}

/** 更新菜单 */
export function updateMenu(id: number, data: MenuUpdateRequest) {
  return request.put<unknown, void>(`/menus/${id}`, data)
}

/** 启用/禁用菜单 */
export function setMenuEnabled(id: number, enabled: boolean) {
  return request.put<unknown, void>(`/menus/${id}/enabled`, null, { params: { enabled } })
}

/** 拖拽排序 */
export function reorderMenu(data: MenuSortRequest) {
  return request.put<unknown, void>('/menus/reorder', data)
}

/** 删除菜单（含子孙节点） */
export function deleteMenu(id: number) {
  return request.delete<unknown, void>(`/menus/${id}`)
}

File: Ivan-fullstack-dev/assets/backend-template/Models/Generate/Menus.cs
using System;

namespace IvanProject.Models;

/// <summary>
/// 菜单实体（统一权限树模型，由 ivan-db-model-gen 根据数据库表自动生成）
/// 对应数据库表：Menus
/// 
/// 存储方案：邻接表 (ParentId) + 路径枚举 (FullPath)
///   - ParentId 保留用于快速插入/移动单节点
///   - FullPath 如 /1/3/7，用于子树查询 LIKE '/1/3/%'
///   - Level 冗余层级，根节点=0
/// 
/// MenuType：1=目录 2=菜单(页面) 3=按钮
/// 按钮节点携带 PermissionCode（权限编码）与 ControllerAction（绑定的后端接口）
/// </summary>
[Serializable()]
public partial class Menus
{
    public Menus() { }

    #region Members

    /// <summary>主键ID</summary>
    public int Id { get; set; }

    /// <summary>菜单名称</summary>
    public string Name { get; set; } = string.Empty;

    /// <summary>路由路径（叶子节点必填，目录节点可为空）</summary>
    public string? Path { get; set; }

    /// <summary>菜单图标</summary>
    public string? Icon { get; set; }

    /// <summary>父级菜单ID（null 为一级菜单/根节点）</summary>
    public int? ParentId { get; set; }

    /// <summary>祖先路径（如 /1/3/7），子树查询加速</summary>
    public string? FullPath { get; set; }

    /// <summary>层级深度（根节点=0，一级子节点=1，以此类推）</summary>
    public int Level { get; set; }

    /// <summary>排序号（同级节点内排序）</summary>
    public int Sort { get; set; }

    /// <summary>是否可见（隐藏菜单仍可访问）</summary>
    public bool? IsVisible { get; set; }

    /// <summary>是否启用（禁用后递归禁用子孙在菜单树的显示）</summary>
    public bool? IsEnabled { get; set; }

    /// <summary>节点类型：1=目录 2=菜单(页面) 3=按钮（统一权限树模型）</summary>
    public int MenuType { get; set; }

    /// <summary>权限编码（按钮节点必填，两段式如 "users:edit"；后端 RequirePerm 与前端 v-perm 共用）</summary>
    public string? PermissionCode { get; set; }

    /// <summary>绑定的后端接口（按钮节点动态维护，如 "UserController.Update"）</summary>
    public string? ControllerAction { get; set; }

    /// <summary>备注</summary>
    public string? Remark { get; set; }

    /// <summary>创建时间</summary>
    public DateTime? CreateTime { get; set; }

    #endregion
}
